src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs

```diff
@@ -3018,5 +3018,524 @@ void F(object o)
                 Diagnostic(ErrorCode.ERR_PatternNullableType, "int?").WithArguments("int").WithLocation(7, 24)
                 );
         }
+
+        [Fact, WorkItem(55668, "https://github.com/dotnet/roslyn/issues/55668")]
+        public void SharedWhenExpression_SwitchExpression()
+        {
+            var source = @"
+M(Position.Last, new Wrap { Sub = new Zero() });
+M(Position.Last, new Wrap { Sub = new One() });
+M(Position.Last, new Wrap { Sub = new Two() });
+M(Position.First, new Wrap { Sub = new Zero() });
+M(Position.Last, new Wrap { Sub = new object() });
+
+static void M(Position position, Wrap wrap)
+{
+    var text = position switch
+    {
+        not Position.First when wrap.Sub is Zero => ""Not First and Zero"",
+        _ when wrap is { Sub: One or Two } => ""One or Two"",
+        Position.First => ""First"",
+        _ => ""Other""
+    };
+
+    System.Console.WriteLine((position, wrap.Sub, text));
+}
+
+enum Position
+{
+    First,
+    Last,
+}
+
+class Zero { }
+class One { }
+class Two { }
+
+class Wrap
+{
+    public object Sub;
+}
+";
+
+            CompileAndVerify(source, expectedOutput: @"
+(Last, Zero, Not First and Zero)
+(Last, One, One or Two)
+(Last, Two, One or Two)
+(First, Zero, First)
+(Last, System.Object, Other)");
+        }
+
+        [Fact, WorkItem(55668, "https://github.com/dotnet/roslyn/issues/55668")]
+        public void SharedWhenExpression_SwitchStatement()
+        {
+            var source = @"
+M(Position.Last, new Wrap { Sub = new Zero() });
+M(Position.Last, new Wrap { Sub = new One() });
+M(Position.Last, new Wrap { Sub = new Two() });
+M(Position.First, new Wrap { Sub = new Zero() });
+M(Position.Last, new Wrap { Sub = new object() });
+
+static void M(Position position, Wrap wrap)
+{
+    string text;
+    switch (position)
+    {
+        case not Position.First when wrap.Sub is Zero: text = ""Not First and Zero""; break;
+        case var _ when wrap is { Sub: One or Two }: text = ""One or Two""; break;
+        case Position.First: text = ""First""; break;
+        default: text = ""Other""; break;
+    }
+
+    System.Console.WriteLine((position, wrap.Sub, text));
+}
+
+enum Position
+{
+    First,
+    Last,
+}
+
+class Zero { }
+class One { }
+class Two { }
+
+class Wrap
+{
+    public object Sub;
+}
+";
+
+            CompileAndVerify(source, expectedOutput: @"
+(Last, Zero, Not First and Zero)
+(Last, One, One or Two)
+(Last, Two, One or Two)
+(First, Zero, First)
+(Last, System.Object, Other)");
+        }
+
+        [Fact, WorkItem(55668, "https://github.com/dotnet/roslyn/issues/55668")]
+        public void SharedWhenExpression_SequencePoints()
+        {
+            var source = @"
+C.M(0, false, false);
+C.M(0, true, false);
+C.M(0, false, true);
+C.M(1, false, false);
+C.M(1, false, true);
+C.M(2, false, false);
+
+public class C
+{
+    public static void M(int i, bool b1, bool b2)
+    {
+        string text;
+        switch (i)
```

Comment:
:memo: To help visualize what's going on, here's the DAG for this example (notice that states 1 and 4 would match if `b2` is true) and the corresponding lowered `switch` statement (with label `<sharedWhenExpression-6>` indicating the shared when expression):
<details><summary>DAG</summary>
```
State 0
    1. [case not 1 when b1:] Not (?DagValueTest(t0 == ConstantValueOne(1: Int32))) WHEN[b1]
    2. [case var _ when b2:] MATCH WHEN[b2]
    3. [case 1:] ?DagValueTest(t0 == ConstantValueOne(1: Int32))
    Test: ?DagValueTest(t0 == ConstantValueOne(1: Int32))
    TrueBranch: 1
    FalseBranch: 3
*State 1 REMAINING t0:[1..1]
    2. [case var _ when b2:] MATCH WHEN[b2]
    3. [case 1:] MATCH
    FalseBranch: 2
*State 2 REMAINING t0:[1..1]
    3. [case 1:] MATCH
*State 3 REMAINING t0:[-2147483648..0],[2..2147483647]
    1. [case not 1 when b1:] MATCH WHEN[b1]
    2. [case var _ when b2:] MATCH WHEN[b2]
    FalseBranch: 4
*State 4 REMAINING t0:[-2147483648..0],[2..2147483647]
    2. [case var _ when b2:] MATCH WHEN[b2]
    FalseBranch: 5
*State 5 FAIL REMAINING t0:[-2147483648..0],[2..2147483647]
```
</details>

<details><summary>Bound tree</summary>
```
block
├─locals: {text}
└─statements
  ├─sequencePointWithSpan
  │ ├─statementOpt
  │ └─span: [213..214)
  ├─sequencePointWithSpan
  │ ├─statementOpt
  │ │ └─block
  │ │   ├─locals: {System.Int32 LoweringTemp.2, System.Int32 SwitchCasePatternMatching.1}
  │ │   └─statements
  │ │     ├─expressionStatement
  │ │     │ └─expression
  │ │     │   └─assignmentOperator
  │ │     │     ├─left
  │ │     │     │ └─local
  │ │     │     │   ├─localSymbol: System.Int32 SwitchCasePatternMatching.1
  │ │     │     │   ├─declarationKind: None
  │ │     │     │   ├─constantValueOpt
  │ │     │     │   ├─isNullableUnknown: False
  │ │     │     │   └─type: int
  │ │     │     ├─right
  │ │     │     │ └─parameter
  │ │     │     │   ├─parameterSymbol: int
  │ │     │     │   └─type: int
  │ │     │     └─type: int
  │ │     ├─sequencePoint
  │ │     │ └─statementOpt
  │ │     ├─block
  │ │     │ └─statements
  │ │     │   ├─conditionalGoto
  │ │     │   │ ├─condition
  │ │     │   │ │ └─binaryOperator
  │ │     │   │ │   ├─operatorKind: IntEqual
  │ │     │   │ │   ├─data
  │ │     │   │ │   ├─resultKind: Viable
  │ │     │   │ │   ├─left
  │ │     │   │ │   │ └─local
  │ │     │   │ │   │   ├─localSymbol: System.Int32 SwitchCasePatternMatching.1
  │ │     │   │ │   │   ├─declarationKind: None
  │ │     │   │ │   │   ├─constantValueOpt
  │ │     │   │ │   │   ├─isNullableUnknown: False
  │ │     │   │ │   │   └─type: int
  │ │     │   │ │   ├─right
  │ │     │   │ │   │ └─literal
  │ │     │   │ │   │   ├─constantValueOpt: ConstantValueOne(1: Int32)
  │ │     │   │ │   │   └─type: int
  │ │     │   │ │   └─type: bool
  │ │     │   │ ├─jumpIfTrue: True
  │ │     │   │ └─label: <dagNode-3>
  │ │     │   └─gotoStatement
  │ │     │     ├─label: <dagNode-4>
  │ │     │     ├─caseExpressionOpt
  │ │     │     └─labelExpressionOpt
  │ │     ├─statementList
  │ │     │ └─statements
  │ │     │   ├─sequencePoint
  │ │     │   │ └─statementOpt
  │ │     │   ├─labelStatement
  │ │     │   │ └─label: <dagNode-4>
  │ │     │   ├─sequencePointWithSpan
  │ │     │   │ ├─statementOpt
  │ │     │   │ │ └─conditionalGoto
  │ │     │   │ │   ├─condition
  │ │     │   │ │   │ └─parameter
  │ │     │   │ │   │   ├─parameterSymbol: bool
  │ │     │   │ │   │   └─type: bool
  │ │     │   │ │   ├─jumpIfTrue: True
  │ │     │   │ │   └─label: case not 1 when b1:
  │ │     │   │ └─span: [292..299)
  │ │     │   ├─sequencePoint
  │ │     │   │ └─statementOpt
  │ │     │   │   └─gotoStatement
  │ │     │   │     ├─label: <dagNode-5>
  │ │     │   │     ├─caseExpressionOpt
  │ │     │   │     └─labelExpressionOpt
  │ │     │   ├─labelStatement
  │ │     │   │ └─label: case not 1 when b1:
  │ │     │   ├─sequencePoint
  │ │     │   │ └─statementOpt
  │ │     │   │   └─expressionStatement
  │ │     │   │     └─expression
  │ │     │   │       └─assignmentOperator
  │ │     │   │         ├─left
  │ │     │   │         │ └─local
  │ │     │   │         │   ├─localSymbol: text
  │ │     │   │         │   ├─declarationKind: None
  │ │     │   │         │   ├─constantValueOpt
  │ │     │   │         │   ├─isNullableUnknown: False
  │ │     │   │         │   └─type: string
  │ │     │   │         ├─right
  │ │     │   │         │ └─literal
  │ │     │   │         │   ├─constantValueOpt: ConstantValueString("b1": String)
  │ │     │   │         │   └─type: string
  │ │     │   │         └─type: string
  │ │     │   └─sequencePoint
  │ │     │     └─statementOpt
  │ │     │       └─gotoStatement
  │ │     │         ├─label: <break-2>
  │ │     │         ├─caseExpressionOpt
  │ │     │         └─labelExpressionOpt
  │ │     ├─statementList
  │ │     │ └─statements
  │ │     │   ├─sequencePoint
  │ │     │   │ └─statementOpt
  │ │     │   ├─labelStatement
  │ │     │   │ └─label: <dagNode-3>
  │ │     │   ├─expressionStatement
  │ │     │   │ └─expression
  │ │     │   │   └─assignmentOperator
  │ │     │   │     ├─left
  │ │     │   │     │ └─local
  │ │     │   │     │   ├─localSymbol: System.Int32 LoweringTemp.2
  │ │     │   │     │   ├─declarationKind: None
  │ │     │   │     │   ├─constantValueOpt
  │ │     │   │     │   ├─isNullableUnknown: False
  │ │     │   │     │   └─type: int
  │ │     │   │     ├─right
  │ │     │   │     │ └─literal
  │ │     │   │     │   ├─constantValueOpt: ConstantValueDefault(0: Int32)
  │ │     │   │     │   └─type: int
  │ │     │   │     └─type: int
  │ │     │   ├─gotoStatement
  │ │     │   │ ├─label: <sharedWhenExpression-6>
  │ │     │   │ ├─caseExpressionOpt
  │ │     │   │ └─labelExpressionOpt
  │ │     │   ├─labelStatement
  │ │     │   │ └─label: <dagNode-5>
  │ │     │   ├─expressionStatement
  │ │     │   │ └─expression
  │ │     │   │   └─assignmentOperator
  │ │     │   │     ├─left
  │ │     │   │     │ └─local
  │ │     │   │     │   ├─localSymbol: System.Int32 LoweringTemp.2
  │ │     │   │     │   ├─declarationKind: None
  │ │     │   │     │   ├─constantValueOpt
  │ │     │   │     │   ├─isNullableUnknown: False
  │ │     │   │     │   └─type: int
  │ │     │   │     ├─right
  │ │     │   │     │ └─literal
  │ │     │   │     │   ├─constantValueOpt: ConstantValueI32(2: Int32)
  │ │     │   │     │   └─type: int
  │ │     │   │     └─type: int
  │ │     │   ├─gotoStatement
  │ │     │   │ ├─label: <sharedWhenExpression-6>
  │ │     │   │ ├─caseExpressionOpt
  │ │     │   │ └─labelExpressionOpt
  │ │     │   ├─labelStatement
  │ │     │   │ └─label: <sharedWhenExpression-6>
  │ │     │   ├─sequencePointWithSpan
  │ │     │   │ ├─statementOpt
  │ │     │   │ │ └─block
  │ │     │   │ │   └─statements
  │ │     │   │ │     ├─conditionalGoto
  │ │     │   │ │     │ ├─condition
  │ │     │   │ │     │ │ └─parameter
  │ │     │   │ │     │ │   ├─parameterSymbol: bool
  │ │     │   │ │     │ │   └─type: bool
  │ │     │   │ │     │ ├─jumpIfTrue: False
  │ │     │   │ │     │ └─label: <afterif-11>
  │ │     │   │ │     ├─block
  │ │     │   │ │     │ └─statements
  │ │     │   │ │     │   ├─switchDispatch
  │ │     │   │ │     │   │ ├─expression
  │ │     │   │ │     │   │ │ └─local
  │ │     │   │ │     │   │ │   ├─localSymbol: System.Int32 LoweringTemp.2
  │ │     │   │ │     │   │ │   ├─declarationKind: None
  │ │     │   │ │     │   │ │   ├─constantValueOpt
  │ │     │   │ │     │   │ │   ├─isNullableUnknown: False
  │ │     │   │ │     │   │ │   └─type: int
  │ │     │   │ │     │   │ ├─cases: {(ConstantValueDefault(0: Int32), <case 0-9>), (ConstantValueI32(2: Int32), <case 2-10>)}
  │ │     │   │ │     │   │ ├─defaultLabel: <break-8>
  │ │     │   │ │     │   │ └─equalityMethod
  │ │     │   │ │     │   ├─labelStatement
  │ │     │   │ │     │   │ └─label: <case 0-9>
  │ │     │   │ │     │   ├─gotoStatement
  │ │     │   │ │     │   │ ├─label: case var _ when b2:
  │ │     │   │ │     │   │ ├─caseExpressionOpt
  │ │     │   │ │     │   │ └─labelExpressionOpt
  │ │     │   │ │     │   ├─labelStatement
  │ │     │   │ │     │   │ └─label: <case 2-10>
  │ │     │   │ │     │   ├─gotoStatement
  │ │     │   │ │     │   │ ├─label: case var _ when b2:
  │ │     │   │ │     │   │ ├─caseExpressionOpt
  │ │     │   │ │     │   │ └─labelExpressionOpt
  │ │     │   │ │     │   └─labelStatement
  │ │     │   │ │     │     └─label: <break-8>
  │ │     │   │ │     └─labelStatement
  │ │     │   │ │       └─label: <afterif-11>
  │ │     │   │ └─span: [379..386)
  │ │     │   ├─sequencePoint
  │ │     │   │ └─statementOpt
  │ │     │   │   └─block
  │ │     │   │     └─statements
  │ │     │   │       ├─switchDispatch
  │ │     │   │       │ ├─expression
  │ │     │   │       │ │ └─local
  │ │     │   │       │ │   ├─localSymbol: System.Int32 LoweringTemp.2
  │ │     │   │       │ │   ├─declarationKind: None
  │ │     │   │       │ │   ├─constantValueOpt
  │ │     │   │       │ │   ├─isNullableUnknown: False
  │ │     │   │       │ │   └─type: int
  │ │     │   │       │ ├─cases: {(ConstantValueDefault(0: Int32), <case 0-13>), (ConstantValueI32(2: Int32), <case 2-14>)}
  │ │     │   │       │ ├─defaultLabel: <break-12>
  │ │     │   │       │ └─equalityMethod
  │ │     │   │       ├─labelStatement
  │ │     │   │       │ └─label: <case 0-13>
  │ │     │   │       ├─gotoStatement
  │ │     │   │       │ ├─label: case 1:
  │ │     │   │       │ ├─caseExpressionOpt
  │ │     │   │       │ └─labelExpressionOpt
  │ │     │   │       ├─labelStatement
  │ │     │   │       │ └─label: <case 2-14>
  │ │     │   │       ├─gotoStatement
  │ │     │   │       │ ├─label: default
  │ │     │   │       │ ├─caseExpressionOpt
  │ │     │   │       │ └─labelExpressionOpt
  │ │     │   │       └─labelStatement
  │ │     │   │         └─label: <break-12>
  │ │     │   ├─labelStatement
  │ │     │   │ └─label: case var _ when b2:
  │ │     │   ├─sequencePoint
  │ │     │   │ └─statementOpt
  │ │     │   │   └─expressionStatement
  │ │     │   │     └─expression
  │ │     │   │       └─assignmentOperator
  │ │     │   │         ├─left
  │ │     │   │         │ └─local
  │ │     │   │         │   ├─localSymbol: text
  │ │     │   │         │   ├─declarationKind: None
  │ │     │   │         │   ├─constantValueOpt
  │ │     │   │         │   ├─isNullableUnknown: False
  │ │     │   │         │   └─type: string
  │ │     │   │         ├─right
  │ │     │   │         │ └─literal
  │ │     │   │         │   ├─constantValueOpt: ConstantValueString("b2": String)
  │ │     │   │         │   └─type: string
  │ │     │   │         └─type: string
  │ │     │   └─sequencePoint
  │ │     │     └─statementOpt
  │ │     │       └─gotoStatement
  │ │     │         ├─label: <break-2>
  │ │     │         ├─caseExpressionOpt
  │ │     │         └─labelExpressionOpt
  │ │     ├─statementList
  │ │     │ └─statements
  │ │     │   ├─sequencePoint
  │ │     │   │ └─statementOpt
  │ │     │   ├─labelStatement
  │ │     │   │ └─label: case 1:
  │ │     │   ├─sequencePoint
  │ │     │   │ └─statementOpt
  │ │     │   │   └─expressionStatement
  │ │     │   │     └─expression
  │ │     │   │       └─assignmentOperator
  │ │     │   │         ├─left
  │ │     │   │         │ └─local
  │ │     │   │         │   ├─localSymbol: text
  │ │     │   │         │   ├─declarationKind: None
  │ │     │   │         │   ├─constantValueOpt
  │ │     │   │         │   ├─isNullableUnknown: False
  │ │     │   │         │   └─type: string
  │ │     │   │         ├─right
  │ │     │   │         │ └─literal
  │ │     │   │         │   ├─constantValueOpt: ConstantValueString("1": String)
  │ │     │   │         │   └─type: string
  │ │     │   │         └─type: string
  │ │     │   └─sequencePoint
  │ │     │     └─statementOpt
  │ │     │       └─gotoStatement
  │ │     │         ├─label: <break-2>
  │ │     │         ├─caseExpressionOpt
  │ │     │         └─labelExpressionOpt
  │ │     ├─statementList
  │ │     │ └─statements
  │ │     │   ├─sequencePoint
  │ │     │   │ └─statementOpt
  │ │     │   ├─labelStatement
  │ │     │   │ └─label: default
  │ │     │   ├─sequencePoint
  │ │     │   │ └─statementOpt
  │ │     │   │   └─expressionStatement
  │ │     │   │     └─expression
  │ │     │   │       └─assignmentOperator
  │ │     │   │         ├─left
  │ │     │   │         │ └─local
  │ │     │   │         │   ├─localSymbol: text
  │ │     │   │         │   ├─declarationKind: None
  │ │     │   │         │   ├─constantValueOpt
  │ │     │   │         │   ├─isNullableUnknown: False
  │ │     │   │         │   └─type: string
  │ │     │   │         ├─right
  │ │     │   │         │ └─literal
  │ │     │   │         │   ├─constantValueOpt: ConstantValueString("default": String)
  │ │     │   │         │   └─type: string
  │ │     │   │         └─type: string
  │ │     │   └─sequencePoint
  │ │     │     └─statementOpt
  │ │     │       └─gotoStatement
  │ │     │         ├─label: <break-2>
  │ │     │         ├─caseExpressionOpt
  │ │     │         └─labelExpressionOpt
  │ │     ├─sequencePoint
  │ │     │ └─statementOpt
  │ │     └─labelStatement
  │ │       └─label: <break-2>
  │ └─span: [246..256)
  ├─sequencePoint
  │ └─statementOpt
  │   └─expressionStatement
  │     └─expression
  │       └─call
  │         ├─receiverOpt
  │         │ └─typeExpression
  │         │   ├─aliasOpt
  │         │   ├─boundContainingTypeOpt
  │         │   ├─boundDimensionsOpt
  │         │   ├─typeWithAnnotations: System.Console
  │         │   └─type: System.Console
  │         ├─method: System.Console.WriteLine(object)
  │         ├─arguments
  │         │ └─conversion
  │         │   ├─operand
  │         │   │ └─objectCreationExpression
  │         │   │   ├─constructor: (int i, bool b1, bool b2, string text).ValueTuple(int, bool, bool, string)
  │         │   │   ├─constructorsGroup: {}
  │         │   │   ├─arguments
  │         │   │   │ ├─parameter
  │         │   │   │ │ ├─parameterSymbol: int
  │         │   │   │ │ └─type: int
  │         │   │   │ ├─parameter
  │         │   │   │ │ ├─parameterSymbol: bool
  │         │   │   │ │ └─type: bool
  │         │   │   │ ├─parameter
  │         │   │   │ │ ├─parameterSymbol: bool
  │         │   │   │ │ └─type: bool
  │         │   │   │ └─local
  │         │   │   │   ├─localSymbol: text
  │         │   │   │   ├─declarationKind: None
  │         │   │   │   ├─constantValueOpt
  │         │   │   │   ├─isNullableUnknown: False
  │         │   │   │   └─type: string
  │         │   │   ├─argumentNamesOpt: (null)
  │         │   │   ├─argumentRefKindsOpt: (null)
  │         │   │   ├─expanded: False
  │         │   │   ├─argsToParamsOpt: (null)
  │         │   │   ├─defaultArguments: Microsoft.CodeAnalysis.BitVector
  │         │   │   ├─constantValueOpt
  │         │   │   ├─initializerExpressionOpt
  │         │   │   ├─wasTargetTyped: False
  │         │   │   └─type: (int i, bool b1, bool b2, string text)
  │         │   ├─conversion: Boxing
  │         │   ├─isBaseConversion: False
  │         │   ├─@checked: False
  │         │   ├─explicitCastInCode: False
  │         │   ├─constantValueOpt
  │         │   ├─conversionGroupOpt
  │         │   ├─originalUserDefinedConversionsOpt: {}
  │         │   └─type: object
  │         ├─argumentNamesOpt: (null)
  │         ├─argumentRefKindsOpt: (null)
  │         ├─isDelegateCall: False
  │         ├─expanded: False
  │         ├─invokedAsExtensionMethod: False
  │         ├─argsToParamsOpt: (null)
  │         ├─defaultArguments: Microsoft.CodeAnalysis.BitVector
  │         ├─resultKind: Viable
  │         ├─originalMethodsOpt: (null)
  │         └─type: void
  └─sequencePointWithSpan
    ├─statementOpt
    │ └─returnStatement
    │   ├─refKind: None
    │   └─expressionOpt
    └─span: [669..670)
```
</details>